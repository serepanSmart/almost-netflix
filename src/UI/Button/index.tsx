import React, { PropsWithChildren } from 'react';
import ButtonContainer  from './styles';

export interface IButtonProps {
  value?: string;
  onClick: (e: React.MouseEvent<HTMLButtonElement>) => void;
  type?: 'button' | 'submit' | 'reset';
  theme?: 'primary' | 'light' | 'reject';
  opacity?: boolean;
  icon?: boolean;
}

const Button: React.FC<PropsWithChildren<IButtonProps>> = ({
  value,
  onClick,
  type,
  theme = 'primary',
  icon,
  children,
}) => {
  return (
    <ButtonContainer
      type={type}
      onClick={onClick}
      theme={theme}
      icon={icon}
    >
      {value}
      {children}
    </ButtonContainer>
  );
};

export default Button;