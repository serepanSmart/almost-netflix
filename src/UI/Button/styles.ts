import styled, { css } from 'styled-components';
import Colors from '../Theme/Colors';
import { IButtonProps } from './index';

type IButtonContainer = Pick<IButtonProps, 'theme' | 'icon'>;

const ButtonContainer = styled.button<IButtonContainer>`
  display: inline-flex;
  justify-content: center;
  padding: ${({ theme }) => theme === 'light'
    ? '10px 20px'
    : '15px 60px'};
  font-size: 20px;
  font-weight: 400;
  text-transform: uppercase;
  line-height: 1;
  background: ${Colors.Scarlet};
  border: 1px solid ${Colors.Scarlet};
  border-radius: 4px;
  color: ${({ theme }) => theme === 'light'
    ? Colors.Scarlet
    : Colors.White};
  cursor: pointer;
  transition: 200ms;
  &:hover {
    background-color: ${Colors.ScarletHover};
  }
  &:active {
    background-color: ${Colors.ScarletFocus};
    color: ${Colors.White};
  }
  ${({ theme }) => theme === 'light' && css`
      background: ${Colors.GreyOpacity};
      border: 1px solid ${Colors.GreyOpacity};
      &:hover {
        background: ${Colors.Scarlet};
        color: ${Colors.White};
        border-color: ${Colors.Scarlet};
      }
  `};
  ${({ theme }) => theme === 'reject' && css`
      background: ${Colors.Tuna};
      color: ${Colors.Scarlet};
      &:hover {
        background: ${Colors.Scarlet};
        color: ${Colors.White};
      }
  `};
  ${({ icon }) => !icon && 'min-width: 210px'};
  ${({ icon }) => icon && css`
      width: 36px;
      height: 36px;
      display: flex;
      align-items: center;
      justify-content: center;
      padding: 6px;
      border-radius: 50%;
      background: ${Colors.Tuna};
      color: ${Colors.White};
      border-color: transparent;
      &:hover {
        background: ${Colors.Grey};
        color: ${Colors.White};
      }
      &:active {
        background: ${Colors.Tuna};
        color: ${Colors.White};
      }
  `};
`;

export default ButtonContainer;