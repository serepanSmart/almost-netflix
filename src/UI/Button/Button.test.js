import { render, fireEvent } from '@testing-library/react';
import Button from '.';
import Colors from '../Theme/Colors';

const handleClick = jest.fn();
const isLoading = false;
const value = 'Button';
const type = 'submit';
const theme = 'reject';
const icon = false;

describe('Button component', () => {
  let component;
  beforeEach(() => {
    component = (props) =>
      render(
        <Button
          value={value}
          isLoading={isLoading}
          theme={theme}
          icon={icon}
          onClick={handleClick}
          type={type}
          {...props}
        />,
      );
  });

  describe('Button render tests', () => {
    test('should define Button component', () => {
      expect(component()).toBeDefined();
    });

    test('should be rendered with value prop', () => {
      const { getByText } = component();
      expect(getByText(value)).toBeInTheDocument();
    });

    test('should match snapshot with default props', () => {
      expect(component()).toMatchSnapshot();
    });

    test('should match with different types', () => {
      const { getByRole } = component({ type: 'reset' });
      expect(getByRole('button')).toBeInTheDocument();
    });

    test('should match with different types', () => {
      const { getByRole } = component({ theme: 'primary' });
      expect(getByRole('button')).toHaveStyle(`background: ${Colors.Scarlet}`);
    });

    test('should match snapshot with different themes', () => {
      const { getByRole } = component();
      expect(getByRole('button')).toBeInTheDocument();
    });

    test('should show loader component when it is loading state', () => {
      const { getByTestId, queryByText } = component({ isLoading: true });
      const loader = getByTestId('bars-loader');
      expect(loader).toBeInTheDocument();
      expect(queryByText(value)).toBeNull();
    });

    test('should show icon without value when it is "icon" prop', () => {
      const { queryByText } = component({ icon: true });
      expect(queryByText(value)).toBeNull();
    });
  });
  describe('Button event handlers', () => {
    test('should call onClick prop when clicked', () => {
      const { getByText } = component();
      fireEvent.click(getByText(value));
      expect(handleClick).toHaveBeenCalledTimes(1);
    });
  });
});
