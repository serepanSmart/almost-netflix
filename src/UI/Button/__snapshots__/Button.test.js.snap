// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Button render tests should match snapshot with default props 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 15px 60px;
  font-size: 20px;
  font-weight: 400;
  text-transform: uppercase;
  line-height: 1;
  background: #f65261;
  border: 1px solid #f65261;
  border-radius: 4px;
  color: #fff;
  cursor: pointer;
  -webkit-transition: 200ms;
  transition: 200ms;
  background: #232323;
  color: #f65261;
  min-width: 210px;
}

.c0:hover {
  background-color: #f65261cc;
}

.c0:active {
  background-color: #90343c;
  color: #fff;
}

.c0:hover {
  background: #f65261;
  color: #fff;
}

<body>
    <div>
      <button
        class="c0"
        type="submit"
        value="Button"
      >
        Button
      </button>
    </div>
  </body>,
  "container": .c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 15px 60px;
  font-size: 20px;
  font-weight: 400;
  text-transform: uppercase;
  line-height: 1;
  background: #f65261;
  border: 1px solid #f65261;
  border-radius: 4px;
  color: #fff;
  cursor: pointer;
  -webkit-transition: 200ms;
  transition: 200ms;
  background: #232323;
  color: #f65261;
  min-width: 210px;
}

.c0:hover {
  background-color: #f65261cc;
}

.c0:active {
  background-color: #90343c;
  color: #fff;
}

.c0:hover {
  background: #f65261;
  color: #fff;
}

<div>
    <button
      class="c0"
      type="submit"
      value="Button"
    >
      Button
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
